examples/service-fastapi/app/security.py

```diff
@@ -0,0 +1,86 @@
+import os
+
+from fastapi import HTTPException
+from fastapi import Security
+from fastapi import status
+from fastapi.security import OAuth2AuthorizationCodeBearer
+from fastapi.security.api_key import APIKeyCookie
+from fastapi.security.api_key import APIKeyQuery
+
+from .client import get_client
+
+### For authenticated endpoints, we want to get auth from one of three ways
+### 1. "token" in the url params
+### 2. "jupyterhub-services" cookie (or config via env)
+### 3. Authorization Bearer header (with oauth to Hub support)
+auth_by_param = APIKeyQuery(name="token", auto_error=False)
+
+COOKIE_NAME = os.getenv("JUPYTERHUB_COOKIE_NAME", "jupyterhub-services")
+auth_by_cookie = APIKeyCookie(name=COOKIE_NAME, auto_error=False)
+
+if "PUBLIC_HOST" in os.environ:
+    ### When running in Docker or maybe other infrastructure,
+    ### JUPYTERHUB_API_URL is "http://jupyterhub" but we need to give
+    ### clients (swagger webpage) a link to the public url
+    auth_url = os.environ["PUBLIC_HOST"] + "/hub/api/oauth2/authorize"
+else:
+    auth_url = os.environ["JUPYTERHUB_API_URL"] + "/oauth2/authorize"
+auth_by_header = OAuth2AuthorizationCodeBearer(
+    authorizationUrl=auth_url, tokenUrl="get_token", auto_error=False
+)
+### ^^ For Oauth in the Swagger webpage, we set the authorizationUrl
+### to the Hub /oauth2/authorize endpoint, so browser does a GET there and
+### receives a 'code' in return.  Then the browser does a POST to our
+### internal /get_token endpoint with that code, and our server subsequently
+### POSTs to Hub /oauth2/token to get/return an acecss_token.
+### The reason for the double POST is that the client (swagger ui) doesn't have
+### the client_secret (JUPYTERHUB_API_TOKEN), but our server does.
+
+### For consideration: build a pydantic User model
```

Comment:
I think that would be cool, just because I'm interested to see what it looks like! Is it easy to allow extra fields with pydantic? Then you can make a subset at least.

Reply:
Pydantic models added and used in `get_current_user` / endpoint routes